Mathlib/GroupTheory/Coxeter/Basic.lean

```diff
@@ -455,9 +495,219 @@ def H₄ : Matrix (Fin 4) (Fin 4) ℕ :=
+    exact Hmul x i h
+  · exact this
+
+/-! ### Homomorphisms from a Coxeter group -/
+
+/-- The proposition that the values of the function `f : B → G` satisfy the Coxeter relations
+corresponding to the matrix `M`. -/
+def IsLiftable {G : Type*} [Monoid G] (M : Matrix B B ℕ) (f : B → G) : Prop :=
+    ∀ i i' : B, (f i * f i') ^ M i i' = 1
+
+private theorem relations_liftable {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f) :
+    ∀ r ∈ CoxeterGroup.Relations.toSet M, (FreeGroup.lift f) r = 1 := by
+  rintro r ⟨⟨i, i'⟩, rfl⟩
+  dsimp [CoxeterGroup.Relations.ofMatrix]
+  rw [map_pow, _root_.map_mul, FreeGroup.lift.of, FreeGroup.lift.of]
+  exact hf i i'
+
+private def groupLift {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  (PresentedGroup.toGroup (relations_liftable hf)).comp cs.mulEquiv.toMonoidHom
+
+private def restrictUnit {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) : B → Gˣ :=
+  fun i ↦ {
+    val := f i,
+    inv := f i,
+    val_inv := pow_one (f i * f i) ▸ (cs.isCoxeter.diagonal i ▸ hf i i),
+    inv_val := pow_one (f i * f i) ▸ (cs.isCoxeter.diagonal i ▸ hf i i),
+  }
+
+/-- Extend the function `f : B → G` to a monoid homomorphism
+`f' : W → G` satisfying `f' (s i) = f i` for all `i`.
+-/
+def lift {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  MonoidHom.comp (Units.coeHom G) (cs.groupLift
+    (show ∀ i i' : B, ((cs.restrictUnit hf) i * (cs.restrictUnit hf) i') ^ M i i' = 1 from
+      fun i i' ↦ Units.ext (hf i i')))
+
+private theorem toMonoidHom_symm (a : PresentedGroup (CoxeterGroup.Relations.toSet M)):
+    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (CoxeterGroup.Relations.toSet M))
+    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
+  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
+  _ = _                                           := by rw [MulEquiv.apply_symm_apply]
+
+theorem lift_apply_simple {G : Type*} [Monoid G] {f : B → G}
+    (hf : IsLiftable M f) (i : B) : cs.lift hf (s i) = f i := by
+  dsimp only [simpleReflection, lift, groupLift, MonoidHom.comp_apply]
+  rw [← MonoidHom.toFun_eq_coe, toMonoidHom_symm cs (PresentedGroup.of i)]
+  simp
+  rfl
+
+/-- If two homomorphisms with domain `W` agree on all simple reflections, then they are equal. -/
+theorem ext_simple {G : Type*} [Monoid G] {φ₁ φ₂ : W →* G} (h : ∀ i : B, φ₁ (s i) = φ₂ (s i)) :
+    φ₁ = φ₂ := by
+  apply MonoidHom.eq_of_eqOn_denseM (cs.submonoid_closure_range_simple)
+  rintro x ⟨i, rfl⟩
+  exact h i
+
+/-! ### Words -/
+/-- The product of the simple reflections of `W` corresponding to the indices in `ω`.-/
+def wordProd (ω : List B) : W := prod (map cs.simpleReflection ω)
+
+local prefix:100 "π" => cs.wordProd
+
+@[simp] theorem wordProd_nil :
+    π [] = 1 := by simp [wordProd]
+
+@[simp] theorem wordProd_cons (i : B) (ω : List B) :
+    π (i :: ω) = s i * π ω := by simp [wordProd]
+
+theorem wordProd_singleton (i : B) :
+    π ([i]) = s i := by simp [wordProd]
```

Comment:
I put it all on one line, but I didn't add the "simp" because `π [i]` unfolds to `π (i :: nil)`, which is not in simp-normal form because of `wordProd_cons`